File: src/main/java/sh/sagan/phobos/builder/SchemBuilder.java
package sh.sagan.phobos.builder;

import com.sk89q.worldedit.EditSession;
import com.sk89q.worldedit.MaxChangedBlocksException;
import com.sk89q.worldedit.WorldEdit;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.extent.clipboard.Clipboard;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormat;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormats;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardReader;
import com.sk89q.worldedit.math.BlockVector3;
import com.sk89q.worldedit.world.World;
import org.bukkit.Location;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;
import sh.sagan.phobos.builder.buildpattern.BuildPattern;
import sh.sagan.phobos.builder.palceeffect.PlaceEffect;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class SchemBuilder {

    private final Clipboard clipboard;
    private final JavaPlugin plugin;

    private int ticksBetweenIterations = 1;
    private boolean ignoreAir = false;
    private BuildPattern buildPattern = (toArrange, clipboard) -> toArrange;
    private PlaceEffect placeEffect = (blockLocation, clipboard) -> {};
    private int blocksPerIteration = Integer.MAX_VALUE;

    private SchemBuilder(JavaPlugin plugin, Clipboard clipboard) {
        this.plugin = plugin;
        this.clipboard = clipboard;
    }

    /**
     * Attempts to construct a new builder instance from the given main plugin and the schematic file. If a builder
     * could not be constructed (invalid file, loading issues), an empty optional is returned.
     *
     * @param plugin The main plugin
     * @param file The schematic file. ({@code Path.of("C:\\schematics\\house.schem").toFile()})
     * @return An optional containing the constructed SchemBuilder is successful, empty otherwise.
     */
    public static Optional<SchemBuilder> from(JavaPlugin plugin, File file) {
        Clipboard clipboard;
        ClipboardFormat format = ClipboardFormats.findByFile(file);
        if (format == null) {
            return Optional.empty();
        }

        try (ClipboardReader reader = format.getReader(new FileInputStream(file))) {
            clipboard = reader.read();
        } catch (IOException e) {
            return Optional.empty();
        }

        return Optional.of(new SchemBuilder(plugin, clipboard));
    }

    /**
     * Sets how many ticks long the period between iterations should be.
     *
     * @param ticks The ticks to wait between iterations
     * @return This builder instance.
     */
    public SchemBuilder ticksBetweenIterations(int ticks) {
        this.ticksBetweenIterations = Math.min(Math.max(1, ticks), 2400);
        return this;
    }

    /**
     * Specifies whether this builder should ignore air when building. When true, this will not alter blocks in the
     * place of air.
     *
     * @param ignoreAir Whether to ignore air blocks when building
     * @return This builder instance.
     */
    public SchemBuilder ignoreAir(boolean ignoreAir) {
        this.ignoreAir = ignoreAir;
        return this;
    }

    /**
     * Sets the build pattern for this builder instance.
     *
     * @param buildPattern The build pattern to use for this builder
     * @return This builder instance.
     */
    public SchemBuilder buildPattern(BuildPattern buildPattern) {
        if (buildPattern != null) {
            this.buildPattern = buildPattern;
        }
        return this;
    }

    /**
     * Sets the place effect for this builder. This effect will be run every time a *single* block is placed.
     *
     * @param placeEffect The place effect to use
     * @return This builder instance.
     */
    public SchemBuilder placeEffect(PlaceEffect placeEffect) {
        if (placeEffect != null) {
            this.placeEffect = placeEffect;
        }
        return this;
    }

    /**
     * Sets the blocks per iteration of this builder. Default is {@code Integer.MAX_VALUE} which will paste all the
     * blocks in one iteration (at once)
     *
     * @param blocksPerIteration The blocks to place per iteration
     * @return This builder instance
     */
    public SchemBuilder blocksPerIteration(int blocksPerIteration) {
        this.blocksPerIteration = Math.max(1, blocksPerIteration);
        return this;
    }

    /**
     * Gets the clipboard object used for this schematic. Use this to alter the clipboard before running the builder
     *
     * @return The clipboard for this loaded schematic.
     */
    public Clipboard getClipboard() {
        return clipboard;
    }

    /**
     * Executes this schem builder, building the schematic at the specified location with all the prior specifications.
     *
     * @param location The location/origin to build this schematic at
     */
    public void buildAt(Location location) {
        if (location.getWorld() == null) {
            return;
        }

        BlockVector3 buildAt = BukkitAdapter.asBlockVector(location);

        new BukkitRunnable(){
            @Override
            public void run() {
                List<Location> toArrange = new ArrayList<>();
                if (ignoreAir) {
                    clipboard.getRegion().forEach(bv -> {
                        if (!clipboard.getBlock(bv).getBlockType().getMaterial().isAir()) {
                            toArrange.add(BukkitAdapter.adapt(location.getWorld(), bv));
                        }
                    });
                } else {
                    clipboard.getRegion().forEach(bv -> toArrange.add(BukkitAdapter.adapt(location.getWorld(), bv)));
                }

                List<BlockVector3> arranged = buildPattern.arrange(toArrange, clipboard).stream()
                        .map(BukkitAdapter::asBlockVector)
                        .collect(Collectors.toList());

                new BukkitRunnable() {
                    @Override
                    public void run() {
                        if (arranged.isEmpty()) {
                            this.cancel();
                            return;
                        }

                        List<BlockVector3> set = new ArrayList<>();
                        if (blocksPerIteration > arranged.size()) {
                            set = new ArrayList<>(arranged);
                            arranged.clear();
                        } else {
                            for (int i = 0; i < blocksPerIteration; i++) {
                                BlockVector3 found;
                                try {
                                    found = arranged.remove(0);
                                } catch (IndexOutOfBoundsException | UnsupportedOperationException exception) {
                                    break;
                                }
                                set.add(found);
                            }
                        }

                        Map<BlockVector3, BlockVector3> toPlace = new HashMap<>();
                        set.forEach(blockVector3 -> {
                            BlockVector3 target = blockVector3.add(buildAt.subtract(clipboard.getOrigin()));
                            toPlace.put(blockVector3, target);
                        });

                        placeBlocks(toPlace, BukkitAdapter.adapt(location.getWorld()));
                    }
                }.runTaskTimer(plugin, 0, ticksBetweenIterations);
            }
        }.runTaskAsynchronously(plugin);
    }

    private void placeBlocks(Map<BlockVector3, BlockVector3> clipboardBlockLocToTarget, World world) {
        try (EditSession editSession = WorldEdit.getInstance().newEditSession(world)) {
            clipboardBlockLocToTarget.forEach((clipboardLoc, target) -> {
                try {
                    editSession.setBlock(target, this.clipboard.getBlock(clipboardLoc));
                    placeEffect.onPlace(BukkitAdapter.adapt(BukkitAdapter.adapt(world), target), this.clipboard);
                } catch (MaxChangedBlocksException e) {
                    e.printStackTrace();
                }
            });
        }
    }
}

File: src/main/java/me/sagan/phobos/builder/SchemBuilder.kt
package me.sagan.phobos.builder

import com.sk89q.worldedit.WorldEdit
import com.sk89q.worldedit.bukkit.BukkitAdapter
import com.sk89q.worldedit.extent.clipboard.Clipboard
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormats
import com.sk89q.worldedit.math.BlockVector3
import me.sagan.phobos.Phobos
import me.sagan.phobos.builder.buildpattern.BuildPattern
import me.sagan.phobos.builder.buildpattern.impl.DefaultBuildPattern
import me.sagan.phobos.builder.effect.BlockPlaceEffect
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.scheduler.BukkitRunnable
import java.io.File
import java.io.FileInputStream

class SchemBuilder(file: File) {

    private var clipboard: Clipboard? = null
    private var ticksBetweenIterations: Long = 2
    private var ignoreAir: Boolean = true
    private var buildPattern: BuildPattern = DefaultBuildPattern()
    private var blockPlaceEffect: BlockPlaceEffect? = null

    init {
        ClipboardFormats.findByFile(file)!!.getReader(FileInputStream(file)).use {
            this.clipboard = it!!.read()
        }
    }

    fun ticksBetweenIterations(ticks: Long): SchemBuilder {
        this.ticksBetweenIterations = ticks
        return this
    }

    fun ignoreAir(ignore: Boolean): SchemBuilder {
        this.ignoreAir = ignore
        return this
    }

    fun buildPattern(buildPattern: BuildPattern): SchemBuilder {
        this.buildPattern = buildPattern
        return this
    }

    fun blockPlaceEffect(blockPlaceEffect: BlockPlaceEffect): SchemBuilder {
        this.blockPlaceEffect = blockPlaceEffect
        return this
    }

    fun buildAt(origin: Location) {

        object : BukkitRunnable() {
            override fun run() {

                // start off by running heavy computations async
                val clipboardActual = clipboard ?: return

                val toSort: MutableList<BlockVector3> =
                        if (ignoreAir) clipboardActual.region.filter { !clipboardActual.getBlock(it).blockType.material.isAir }.toMutableList()
                        else clipboardActual.region.toMutableList()

                val sorted: List<List<BlockVector3>> = buildPattern.sort(toSort, clipboardActual)

                // hop back on sync to place the blocks
                object : BukkitRunnable() {
                    var i: Int = 0
                    override fun run() {
                        if (i >= sorted.size) {
                            this.cancel()
                            return
                        }

                        val toPlace: MutableMap<BlockVector3, BlockVector3> = mutableMapOf()
                        for (blockVector3 in sorted[i]) {
                            val actualLoc: BlockVector3 = blockVector3.add(BukkitAdapter.asBlockVector(origin).subtract(clipboardActual.origin))
                            toPlace[blockVector3] = actualLoc

                            blockPlaceEffect?.onPlace(Location(origin.world, actualLoc.x.toDouble(), actualLoc.y.toDouble(), actualLoc.z.toDouble()))
                        }

                        placeBlocks(toPlace, origin.world!!)
                        i++
                    }
                }.runTaskTimer(Phobos.instance!!, 0, ticksBetweenIterations)
            }
        }.runTaskAsynchronously(Phobos.instance!!)
    }

    private fun placeBlocks(clipboardToActual: Map<BlockVector3, BlockVector3>, world: World) {
        val clipboardActual = this.clipboard ?: return
        WorldEdit.getInstance().editSessionFactory.getEditSession(BukkitAdapter.adapt(world), -1).use {
            for ((clipboardLoc: BlockVector3?, actual: BlockVector3?) in clipboardToActual) {
                it.setBlock(actual, clipboardActual.getBlock(clipboardLoc))
            }
        }
    }
}